[Merged by Bors] - refactor: replace Bitvec with Std.BitVec

mhk119 opened the pull request:
re-adding definitions and theorems about bitvectors that mathlib has for `Bitvec` to the new `Std.BitVec` type, and getting rid of the former.

In std, the choice was made to define ult (unsigned less than) and related comparisons as `Bool`-valued, where the corresponding defs for `Bitvec` in mathlib are `Prop`-valued. Std does not have definitions for the respective greater-equal or greater-than comparisons, so we add those here, but we choose to be consistent with std and make them `Bool`-valued as well, breaking with the old API. 

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: 
-->
- [x] depends on: #8598

EDIT (alexkeizer): 
Since opening this PR, substantial parts of the proposed refactor have been upstreamed to std.
This PR has thus changed its goal from:

> Redefine `BitVec w` as `Fin (2^w)` instead of `Vector Bool w` because the `Fin` representation is computationally more efficient and allows users to reuse existing theorems on `Fin` or `Nat` easily.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

alexkeizer commented:
@eric-wieser could you clarify why you labelled this after-port?

A [search](https://github.com/search?q=repo%3Aleanprover-community%2Fmathlib%20bitvec&type=code) of mathlib3 shows that the only downstream dependent file (control.random) is already ported, so there should be no problems with changing the representation.

eric-wieser commented:
"after-port" means "this might cause trouble for downstream projects". At any rate, the "port-complete" tag should be made in the next few days once we finish the `align_import` stuff.

alexkeizer commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,355 +1,239 @@
 
-private def repr {n : Nat} : Bitvec n → String
-  | ⟨bs, _⟩ => "0b" ++ (bs.map fun b : Bool => if b then '1' else '0').asString
+theorem BV_add {x y : BitVec w}: bitadd x.val y.val w = (x + y).val := by
```
I'm not 100% sure we want to have `bitadd` and related definitions on `Nat`; it seems to overlap quite a bit with the `Fin n -> Bool` alternative representation that was discussed on Zulip.

I would suggest limiting this PR to just the basic definitions on `Bitvec`tors, and removing all the extra definitions from `Nat.Bitwise`.

mhk119 replied:
Ok, I made that change. Now everything included is just necessary for definitions on ```BitVec```.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,355 +1,239 @@
-section Conversion
+-- We inherit `Fin` implementations when fast but write mod/div
+-- ourselves to avoid the extra modulo operation.
+protected def add (x y : BitVec w) : BitVec w := Fin.add x y
```
I don't think there's anything interesting to read here, so the whitespace-less version seems fine to me

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,355 +1,239 @@
+protected def le (x y : BitVec w) : Bool :=
+  x.val ≤ y.val
+
+instance : Add (BitVec w) := ⟨BitVec.add⟩
```
`instance : Ring (BitVec w) := inferInstanceAs (Ring(Fin _))` should be enough to replace all of them.

mhk119 replied:
It said failed to synthesize instance `Ring (Fin (2 ^ w))`

alexkeizer replied:
Try to import `Mathlib.Data.ZMod.Defs`?

eric-wieser commented on Mathlib/Init/Data/Nat/Bitwise.lean:
```diff
@@ -221,6 +227,11 @@ theorem shiftr_zero : ∀ n, shiftr 0 n = 0 := by
     rw [shiftr, div2, IH]
     rfl
 
+theorem shiftRight_eq_shiftr: n >>> m = Nat.shiftr n m := by
+  simp only [HShiftRight.hShiftRight, ShiftRight.shiftRight]
+  induction' m with m hm generalizing n
+  <;> simp [Nat.shiftr, Nat.shiftRight, Nat.div2_val, *]
```
This looks to me like an indication that we should just delete `shiftr`, we don't need two ways to spell the same function. Can you make a separate PR to do that (and the same for `shiftLeft`?

mhk119 replied:
Sure!

mhk119 replied:
Ok [https://github.com/leanprover-community/mathlib4/pull/6356](here) it is.

eric-wieser replied:
This can now be deleted.

alexkeizer commented:
@mhk119 I'm curious what the status of this PR is?
There was some back and forth between @eric-wieser's review and your code changes. Have all the review concerns now been addressed, or have there been significant other changes so that another review is needed?

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -1,17 +1,22 @@
-Authors: Markus Himmel
+Authors: Markus Himmel, Harun Khan, Abdalrhman M Mohamed
+
+! This file was ported from Lean 3 source module data.nat.bitwise
+! leanprover-community/mathlib commit 6afc9b06856ad973f6a2619e3e8a0a8d537a58f2
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
```
This is the old header format which we no longer use

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,178 @@
 /-
-Copyright (c) 2015 Joe Hendrix. All rights reserved.
+Copyright (c) 2022 by the authors listed in the file AUTHORS and their
+institutional affiliations. All rights reserved.
```
What happened here?

eric-wieser commented:
Please merge master

eric-wieser commented on Mathlib/Data/Bool/Basic.lean:
```diff
@@ -372,13 +375,15 @@ def toNat (b : Bool) : Nat :=
 
 theorem ofNat_le_ofNat {n m : Nat} (h : n ≤ m) : ofNat n ≤ ofNat m := by
-  simp only [ofNat, ne_eq, _root_.decide_not]
+  simp only [ofNat, ne_eq, _root_.decide_not];
```
Can you undo the new semicolons in this file?

mhk119 replied:
Just one semicolon?

eric-wieser commented on Mathlib/Data/Bool/Basic.lean:
```diff
@@ -2,13 +2,16 @@
 import Mathlib.Init.Data.Nat.Lemmas
 import Mathlib.Init.Function
 
-#align_import data.bool.basic from "leanprover-community/mathlib"@"c4658a649d216f57e99621708b09dcb3dcccbd23"
```
Please don't convert these to old headers like this; I would guess you did a bad merge at some point?

mhk119 replied:
Yeah. I think that's due to some merge.

eric-wieser replied:
I've unresolved this because the offending change is still here.

mhk119 replied:
I see this header is also in master.

mhk119 replied:
I see the old header in this file in the master branch 🤔.

mhk119 commented:
> @mhk119 I'm curious what the status of this PR is? There was some back and forth between @eric-wieser's review and your code changes. Have all the review concerns now been addressed, or have there been significant other changes so that another review is needed?

Yes, we just PR'ed [6356](https://github.com/leanprover-community/mathlib4/pull/6356) which means we can now continue working on the new changes.

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -171,6 +231,52 @@ theorem bitwise'_comm {f : Bool → Bool → Bool} (hf : ∀ b b', f b b' = f b'
     _ = swap (bitwise' f) := bitwise'_swap hf'
 #align nat.bitwise_comm Nat.bitwise'_comm
 
+lemma bitwise_eq_bitwise' (f: Bool → Bool → Bool ) (h: f false false = false) :
```
Did you have a counterexample that demonstrated why we need `h`? This was discussed on Zulip, right?

mhk119 replied:
Yes. From what I understood, we don't need ```h```. However, without it, it's hard to prove...

eric-wieser replied:
I think we should:
* Prove that they are equal
* Remove one of them, unless there's a good reason to have both definitions.

alexkeizer replied:
I've been trying to prove this lemma without `h`, but it is indeed quite hard.

Both `bitwise` and `bitwise'` already exist in master, and a decent number of existing `bitwise'` theorems carry a similar assumption, so I'd argue that this current PR does not worsen the current status quo and I don't think we should block it on this issue.

(For future reference: the zulip thread mentioned is: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Merging.20bitwise.20and.20bitwise')

mhk119 replied:
Yes, I'm trying to prove this too and it is indeed quite challenging.

ChrisHughes24 replied:
Can I make the following suggestions for how to prove it. The issue seems to boil down to the fact that in a sense, `binaryRec` defines the case `n=0` twice, because `bit false 0 = 0`.

I think there should be a version of `binaryRec_eq` with the different hypothesis `n = 0 -> b = true`

```
theorem binaryRec_eq' {C : Nat → Sort u} {z : C 0} {f : ∀ b n, C n → C (bit b n)}
    (b n) (hbn : n = 0 → b = true) : binaryRec z f (bit b n) = f b n (binaryRec z f n)
```

Then the theorem `bitwise'_bit` can be stated as 
```
theorem bitwise'_bit' {f : Bool → Bool → Bool} (a m b n) 
    (ham : m = 0 → a = true) (hbn : n = 0 → b = true) :
    bitwise' f (bit a m) (bit b n) = bit (f a b) (bitwise' f m n) :=
```
And theorems could even be proven with a stronger version of `binaryRec`.
```
def binaryRec' {C : Nat → Sort u} (z : C 0) (f : ∀ b n, (n = 0 → b = true) → C n → C (bit b n)) : ∀ n, C n :=
```

alexkeizer replied:
Thanks! It turns out `Nat.binaryRec_eq'` and `binaryRec'` already exist in `Data.Nat.Bits`. With the former, I finished and pushed a proof of `bitwise_eq_bitwise'` that doesn't need the extra assumption

eric-wieser commented on Mathlib/Init/Data/Nat/Bitwise.lean:
All the changes to this file look accidental to me.

eric-wieser replied:
This comment still applies, since the file still has changes in the PR

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -1,17 +1,17 @@
 import Mathlib.Data.Nat.Order.Lemmas
 import Mathlib.Tactic.Linarith
 
-#align_import data.nat.bitwise from "leanprover-community/mathlib"@"6afc9b06856ad973f6a2619e3e8a0a8d537a58f2"
```
Accidental deletion?

tobiasgrosser replied:
Is this the only remaining issue? Or is there sth else to be done in this PR?

eric-wieser replied:
I would prefer to see a separate PR that addresses https://github.com/leanprover-community/mathlib4/pull/5920#discussion_r1310496092 first; though maybe another reviewer can give a second opinion.

eric-wieser replied:
I've marked this unresolved again because this line is still deleted and shouldn't be.

tobiasgrosser replied:
This has now been reverted.

eric-wieser commented on Mathlib/Init/Data/Nat/Bitwise.lean:
```diff
@@ -180,21 +180,21 @@ theorem bit_zero : bit false 0 = 0 :=
  and adds the bit `b` as the least significant bit each time.
  Returns the corresponding natural number-/
-def shiftl' (b : Bool) (m : ℕ) : ℕ → ℕ
+def shiftLeft' (b : Bool) (m : ℕ) : ℕ → ℕ
```
Can you merge master? I already did this in #6788

eric-wieser commented:
Sorry for the rather piecewise reviews.

This is quite a big change overall; can you write a PR description that summarizes the change and explains why the new definition is better than the old definition?

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,189 @@
+instance : Inhabited (BitVec w) := ⟨BitVec.zero w⟩
 
-/-! ### Comparison operators -/
+instance : OfNat (BitVec w) (nat_lit 0) :=
```
I think you should be using the `Zero` typeclass here.

alexkeizer commented:
> Sorry for the rather piecewise reviews.
> 
> This is quite a big change overall; can you write a PR description that summarizes the change and explains why the new definition is better than the old definition?

For future reference: this refactor was also discussed in a [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Proposed.20refactor.20of.20Bitvec)

alexkeizer commented:
@mhk119 I think all the comments raised in the review have now been addressed, right? You might want to change the label from `awaiting-author` to `awaiting-review`

tobiasgrosser commented:
@eric-wieser, I addressed your comment about the accidentally dropped `#align_import` and would appreciate another pass.

(We are planning to use and extend this bitvector theory, and slowly reach a point where having this in mathlib would help us avoid doing redundant work out-of-tree)

linesthatinterlace commented:
For what it's worth, I do not think this is the right approach. I appreciate that the plan is to re-enable reasoning about Bitvec as a vector of bits, but I think prioritising the arithmetic representation in this way is a misguided idea. I have been asked by @ericrbg to give an extension of my thoughts (some of which I expressed further in that Zulip thread).

Firstly, for context, I've recently been working with Fin(2^w) considered as a bitvector in some work of mine: the key stuff I have extracted out here. https://gist.github.com/linesthatinterlace/4843a288498cc8f258212e864b8b2ff6

In particular, being able to easily "pull out" one of the bits of Fin(2^(w + 1)) is a really nice feature. Notice I don't use the current Bitvec representation (which I also think is bad): but I do use finFunctionFinEquiv to convert between Fin(2^w) and Fin(w) -> Fin 2. I think on some level any approach which does not use this equivalence is missing a trick, because it already does so much work for you.

Essentially we have two reasonable definitions of Bitvec w, right? `Fin(w) -> Fin 2` and `Fin(2^w)`. You could also have `Fin(w) -> Bool`, I slightly prefer that for various reasons, but that is much of a muchness.

The idea behind this PR is that arithmetic operations on Bitvec w should use the Fin(2^w) representation, because that is much faster/you are using native Nat addition. I 100% agree with this.

However, bitwise operations, or bit-manipulation operations - such as complement, or and - should not be done, as here, by converting to the underlying nat values, and *then* doing the bitwise operations on Nat to find the bitwise-and (which is effectively I think what is happening). Because this loses the fact that the numbers have exactly w bits - bitwise-and for Nat is for nats where we don't specify an underlying bitlength, whereas here by definition we do know this. So I cannot believe that the approach here is the correct one. *Clearly* Fin w -> Fin 2 is the better representation here (or, actually, Fin w -> Bool: again, I like this more, but that's a different issue). You also want to be able to do things like "pull out a bit from the bitwise operation, and tweak that bit based on the number that remains" (this is what I need to do in my work). If you want to do this over Fin(2^w), really you are going to have to convert to Fin w -> Fin 2 first.

Essentially you have two types: "Bitvec w as an arithmetic animal" and "Bitvec w as a vector animal". Both of these have a set of operations and a representation that makes them more efficient. Actually, they also both have an order (former the arithmetic order, the latter the lexographic order), and these orders are compatible. That's sort of I think the whole ballgame.

The current approach for Bitvec prioritises the vector approach at the cost of the arithmetic approach (albeit also doing the vector approach in a bad way - tuples are preferable I think). This PR would prioritise the arithmetic approach, but I think the choices it makes with reference to bits are bad, even with the caveat that future PRs are intended to bridge the gap: at the heart the approach ignores the advantages that fixed-width bitwise or bit-twiddling operations give you. I do not think it is the right way to go.

One possibility would be that you have Bitvec w (Fin w -> Fin 2 or Fin w -> Bool) and FixedWidth w (Fin (2^w)), with the obvious equivalence between them, and define the operations that are natural on each of them in the sensible way, and the operations that aren't natural (e.g. addition on Bitvec w) you define via conjugation by the equivalence. But you can then also show that this is equivalent to the things that ought to be true about it, if you like. And if you really wanted to be neutral you could have a third type which contained both the BitVec and FixedWidth representations of a number and a proof that they were equivalent, and simply do the required operations on that.

Anyway: these are just my thoughts as I was alerted to this by @ericrbg: but speaking as someone who has been working with the bit-representation of Fin(2^w) recently, I think this PR completely ignores things like `finFunctionFinEquiv` and is worse for it, and I do not believe it is the right approach.

linesthatinterlace commented:
In short, I suppose: I agree with this. https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Proposed.20refactor.20of.20Bitvec/near/376094740, and I think the discussion there captured much of what I've tried to say above, and contains a lot more nuance than the poll which I think doesn't fairly present the options.

ChrisHughes24 commented:
I agree with @linesthatinterlace that the dependence on `Nat.bitwise` is bad, mainly for the reason that `Nat.bitwise` is wrong whenever `f false false = true`.  It is probably a faster way to get the definition of `and` and `or` etc. on Bitvectors, but very quickly we will want a generic bitwise on Bitvector, and this cannot depend on `Nat.bitwise` because that definition is wrong when `f false false = true`. So I think all bitwise stuff should probably depend on a generic `bitwise` on bitvectors.

I disagree with the proposal for two named types, I think we want one `Bitvector`, but definitely we want functions `Bitvec n -> Fin n -> Bool` and `Bitvec n -> Vector n Bool`, but we shouldn't have any type alias for these. I would keep the `Fin (2 ^ n)` representation, but have enough API that the user can induct on it like a vector or use it like a function to bool if they want.

linesthatinterlace commented:
I see what you're saying, Chris: I actually think we want *equivalences* rather than functions but same difference. I don't wholly disagree (though if we only have one type synonym, I think we should called Fin(2^m) FixedWidth and NOT Bitvector: I don't think we should call something Bitvector when it isn't vectorish in shape).

One reason to favour named types, as a counter proposal, is that it can help with poking the typechecker to unify stuff. For example, sometimes one might have something of type Fin (1 + 1) -> Bool, and you want to treat this as (what I will synonymise as) Fin 2 -> Bool. Actually Lean can do this with no effort, it can unify the types because 1 + 1 is defeq to 2, but if everything is written explictly it can have issues.

(There are similar issues with Fin (2^0) and Fin 1, incidentally, and Fin (2^1) and Fin 2. It's another reason why having a type synonym there can be useful.)

It may be that there's better ways to solve this type of issue but I have noticed that sometimes giving a name to a complicated type involving Fin (expression) helps with type unification.

linesthatinterlace commented:
And, yes, frankly Nat.bitwise is *also* not ideal, and it would be good to see it nixed and integrated with this, rather than used as a foundation...

(Every Nat can be coerced to a suitable Bitvector/FixedWith by using, I think it's called nextPowerOfTwo?)

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,189 @@
 
+/-!
+## Arithmetic
+We inherit `Fin` implementations when fast but write mod/div
+ourselves to avoid the extra modulo operation.
+-/
+/-- The mod of two bitvectors. -/
+protected def mod (x y : BitVec w) : BitVec w :=
+  ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.isLt⟩
+
+/-- Dividing two bitvectors. -/
+protected def div (x y : BitVec w) : BitVec w :=
+  ⟨x.val / y.val, Nat.lt_of_le_of_lt (Nat.div_le_self _ _) x.isLt⟩
```
Can this extra mod operation just be removed from the existing definitions?

mhk119 replied:
I believe it is indeed unnecessary in the `Fin` definition.

mhk119 replied:
Actually, there is a comment in the file `Init.Data.Fin.Basic` saying 
```
/-!
Remark: mod/div/modn/land/lor can be defined without using (% n), but
we are trying to minimize the number of Nat theorems
needed to boostrap Lean.
-/
```
Hence, I suppose we leave the `Fin` definitions as is.

eric-wieser replied:
If this is just about runtime performance, I believe the thing to do is write a `csimp` lemma that rewrites the slow implementations into fast implementations; then:
* you don't actually need a new implementation
* the implementation ends up being fast for `Fin` too, not just `BitVec`

eric-wieser replied:
(please do this in a standalone PR!)

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+
+/-- `x < y` as natural numbers if and only if `x < y` as `BitVec w`. -/
+@[norm_cast]
+theorem val_bitvec_lt {x y : BitVec w} : x.val < y.val ↔ x < y := by
```
`val_lt_val`

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+
+/-- `x ≠ y` as natural numbers if and only if `x != y` as `BitVec w`. -/
+@[norm_cast]
+theorem val_bitvec_bne {x y : BitVec w} : x.val ≠ y.val ↔ x != y := by
```
`val_ne_val_iff_bne`

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+theorem val_bitvec_bne {x y : BitVec w} : x.val ≠ y.val ↔ x != y := by
+  simp [bne, val_bitvec_eq]
+
+lemma testBit_eq_ofNat {k} {x : BitVec w} :
```
I think shift should be mentioned in the name of this lemma

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+  simp only [BitVec.ofNat, Fin.ofNat', testBit, mod_two_of_bodd, pow_one]
+  aesop
+
+lemma val_to_ofNat {m} (h : m < 2^w) : (BitVec.ofNat w m).val = m := by
```
`val_ofNat`

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+lemma val_to_ofNat {m} (h : m < 2^w) : (BitVec.ofNat w m).val = m := by
+  simp [BitVec.ofNat, Fin.ofNat', mod_eq_of_lt h]
+
+lemma ofNat_to_val (x : BitVec w) : BitVec.ofNat w x.val = x := by
```
`ofNat_val`

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+lemma ofNat_to_val' (x : BitVec w) (h : v = w):
+  HEq x (BitVec.ofNat v x.val) := h ▸ heq_of_eq (ofNat_to_val x).symm
+
+theorem concat_ext {x : BitVec w} {y : BitVec v} :
```
This is about `append` not `concat` right? And it's not an extensionality theorem. It should have a name that has more to do with the statement.

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+  (x ++ y).val = x.val <<< v ||| y.val := rfl
+
+theorem extract_ext {i j} {x : BitVec w} :
+  (extract i j x).val = x.val / 2 ^ j % (2 ^ (i - j + 1)) := by
```
`val_extract_eq_div_pow_mod`

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,68 @@
+  (extract i j x).val = x.val / 2 ^ j % (2 ^ (i - j + 1)) := by
+  simp [extract, BitVec.ofNat, Fin.ofNat', shiftRight_eq_div_pow]
+
+theorem get_eq_testBit {x : BitVec w} {i} : x.get i = x.val.testBit i := by
```
Why do we have three ways of saying the same thing?

alexkeizer commented:
> I agree with @linesthatinterlace that the dependence on Nat.bitwise is bad, mainly for the reason that Nat.bitwise is wrong whenever f false false = true. It is probably a faster way to get the definition of and and or etc. on Bitvectors, but very quickly we will want a generic bitwise on Bitvector, and this cannot depend on Nat.bitwise because that definition is wrong when f false false = true. So I think all bitwise stuff should probably depend on a generic bitwise on bitvectors.

Although I agree `Nat.bitwise` is a bit of a footgun because of the mentioned reasons, I don't really agree with the conclusion. We aren't so much choosing to depend on `Nat.bitwise` as re-using bitwise operations on `Nat`, which happen to be modeled in terms of `Nat.bitwise`. Just like for arithmetic, doing this gives us bitwise `and`/`or`/`not` in a single instruction (assuming `n <= 64`).

I don't agree that the difference between `Fin 2^n` and `Fin n -> Bool`/`Vector Bool n` lies in arithmetic versus bitwise perspectives. In my opinion this is about whether you want optimized operations (both arithmetic *and* bitwise), or whether you want easy to reason about operations.

If we define a `BitVec.bitwise`, I wholeheartedly agree we should stay away from `Nat.bitwise`. I just don't think we should use this `BitVec.bitwise` to define bitwise `and` (and co.) --- we should just prove a theorem allowing us to rewrite `and` in terms of `BitVec.bitwise` for reasoning.

mhk119 commented:
Thanks @linesthatinterlace for your comment. Right, the idea is to define Bitvector operations in terms of operations on underlying natural numbers because it is much faster. In light of this, there are some operations that clearly have a much nicer elegant definition in terms of naturals (such as addition and multiplication). On the other hand, it's not clear that defining bitwise operations such as and/or is nicer this way (in fact, it probably isn't but it's still faster :)). My position is to define all of the fundamental operations in terms of underlying naturals for efficiency. And a ```BitVec.bitwise``` could be used for less fundamental operations I suppose.

Then with these definitions, we will prove that they are equivalent to the corresponding vector-like definitions (in future PRs). It looks like ```finFunctionFinEquiv``` would be useful here (I was unaware of this so thanks for bringing this up). Personally, just having a function ```BitVec n -> Fin n -> Bool``` is cleaner than a named type.

linesthatinterlace commented:
I think you've persuaded me that this is the right approach. I am a bit surprised that, e.g., finding the complement by calculating 0 - (x+ 1) is the performant way of doing it, but if it is, it is. It would be good to see metrics on this: if we are going to have a Bitwise type that prioritises performance in all things, let's make sure we definitely are making the performant choices.

Agree that we can then simply have a function to Fin n -> Bool. Actually I believe we should definitely have a function both ways, i.e. we should define an equiv (which will probably just be finFunctionFinEquiv in a hat): it'll just be a lot easier to work with because sometimes you'll want to pass into the one representation and pass out again (see my getBitResiduum above for an example of what I mean - actually there probably are more performant ways of doing the kind of twiddle I do there but I'm not sure I know how to prove they work...)

What I want to make sure we're doing is that everything lies flat and we don't end up proving the same thing twice separately, or have parallel theories that don't quite work. The current PR does still "tastes messy" to me in this sense. In my experience, tiny choices with regards to things like "how does this actually get bits" can make a big difference.

If you really think this is the best way, however, and we've done the analysis to prove it, then there's no reason not to use this approach.

tobiasgrosser commented:
I am not a theorem proving expert, but using native integers on a CPU, we can compute the complement for up to 64 bit in three cycles latency and probably even higher throughput. Any representation that uses 'lists of sth' will be in the 10th to 100s of cycles, cannot be vectorized, .... So with respect to efficiency, I think there is clear agreement that the proposed representation here is best. It also seems we have plans forward to develop an effective proof infrastructure to make this high-performance implementation good to reason about.

linesthatinterlace commented:
I'm very excited, then. This has a lot of great potential.

What are the chances by the way that you could do something like synthesizing a proof for facts about BVs using a suitable SMT solver?

mhk119 commented:
> I'm very excited, then. This has a lot of great potential.
> 
> What are the chances by the way that you could do something like synthesizing a proof for facts about BVs using a suitable SMT solver?

Yes that is the idea. This PR stemmed from our project [lean-smt](https://github.com/mhk119/lean-smt/blob/bitvec/Smt/Data/BitVec.lean) where we are making a Lean tactic that can call an smt-solver to close goals. Really exciting!

linesthatinterlace commented:
I did see some comments that made me wonder if you were thinking along those lines. Very cool.

tobiasgrosser commented:
Indeed. @mhk119's work will make this so much more useful. We are working on using the same representation to optimize programs. This PR is foundational to several efforts in this area, and agreeing on a joint representation will enable us to collaborate a lot more effectively. It is nice to see more and more convergence here. I am looking forward to seeing the last open questions resolved.

tobiasgrosser commented:
One of our first steps is to proof LLVM/MLIR peephole rewrites with Lean. We have plenty of Bitvector level theorems to be proven correct: https://github.com/opencompl/ssa/blob/main/SSA/Projects/InstCombine/AliveStatements.lean and these are used in llvm.org and mlir.llvm.org to optimize the speed of programs. Coincidentally, @bollu and @hargoniX are also working on an LLVM backend for Lean itself (https://github.com/leanprover/lean4/milestone/3). 

Anyway, let me not distract from this PR further. It seems we are on a good track here.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+instance : Ring (BitVec w) := inferInstanceAs (Ring (Fin _))
+instance : Mod (BitVec w) := inferInstanceAs (Mod (Fin _))
+instance : Div (BitVec w) := inferInstanceAs (Div (Fin _))
+instance : LT (BitVec w)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec w)  := ⟨fun x y => BitVec.le x y⟩
```
Why not inherit the `LinearOrder` instance from `Fin` here?

Perhaps a better question: is there anything you *don't* want to inherit from `Fin`? If not, making `BitVec` an `abbrev` would save you a lot of work.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
 
-instance {n} {x y : Bitvec n} : Decidable (Bitvec.Ugt x y) :=
-  decEq _ _
+/-- Less than for two bitvectors. -/
+protected def lt (x y : BitVec w) : Bool :=
+  x.val < y.val
```
Note this should have been a `Prop` not a `Bool`, though my comment below advocates deleting this anyway.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+      rw [Nat.mul_add, Nat.add_comm, Nat.mul_one]
+    hEq ▸ (x ++ replicate n x)
+
+/-- Pad a bitvector by the zero bitvector of length `i`. -/
```
Is this padding the MSB or the LSB?

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
 
-instance : Sub (Bitvec n) :=
-  ⟨Bitvec.sub⟩
+/-- The zero bitvector. -/
+protected def zero (w : Nat) : BitVec w :=
+  ⟨0, Nat.two_pow_pos w⟩
```
There's no point in having this when you have `0 : BitVec w`

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
 
-instance : Neg (Bitvec n) :=
-  ⟨Bitvec.neg⟩
+/-- The bitvector `n mod 2^w`. -/
+protected def ofNat (w : Nat) (n : Nat) : BitVec w :=
+  Fin.ofNat' n (Nat.two_pow_pos w)
```
Similarly this can be removed, it's implied by the `Ring` instance below.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Concatenation of two bitvectors. -/
```
Please indicate which argument contains the MSBs and which is the LSBs; perhaps by renaming the arguments!

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+instance : HAppend (BitVec w) (BitVec v) (BitVec (w + v)) := ⟨BitVec.append⟩
+
+/-- Extract the bitvector between indices i and j. -/
+@[simp]
+def extract (i j) (x : BitVec w) : BitVec (i - j + 1) :=
+  BitVec.ofNat _ (x.val >>> j)
```
I would argue this should be
```suggestion
@[simp]
def extract (i j) (x : BitVec w) : BitVec (i - j) :=
  BitVec.ofNat _ (x.val >>> j)
```
that is, `extract 0 8 x` obtains the lowest 8 bits, and `extract i j x ++ extract j k x = extract i k x` (or perhaps the other way around, since you didn't document `++` clearly enough!)

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+
+/-- Keep the `v` least significant bits of a bitvector. -/
+def shrink (v) (x : BitVec w) : BitVec v :=
+  if hZero : 0 < v then
+    have hEq : v - 1 + 0 + 1 = v := by
+      rw [Nat.add_zero]
+      exact Nat.sub_add_cancel hZero
+    hEq ▸ x.extract (v - 1) 0
+  else 0
```
With my suggested change to `extract` this becomes
```suggestion
  x.extract v 0
```
which probably indicates you don't need the definition at all.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+def get (x : BitVec w) (i : Nat) : Bool :=
+  x.extract i i != 0
+
+/-- View a bitvector as a list of booleans. -/
+def bbT (bs : List Bool) : BitVec bs.length :=
```
This docstring is backwards

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,177 @@
+-/
 
-instance {n} : ShiftRight (Bitvec n) := ⟨fun x y => x >>> y.toNat⟩
+/-- The complement of a bitvector. -/
+protected def complement (x : BitVec w) : BitVec w :=
+  0 - (x + .ofNat w 1)
+
+/-- The and of two bitvectors. -/
+protected def and (x y : BitVec w) : BitVec w :=
+  ⟨x.val &&& y.val, by simp [HAnd.hAnd, AndOp.and, Nat.land, Nat.bitwise_lt]⟩
+
+/-- The or of two bitvectors. -/
+protected def or (x y : BitVec w) : BitVec w :=
+  ⟨x.val ||| y.val, by simp [HOr.hOr, OrOp.or, Nat.lor, Nat.bitwise_lt]⟩
+
+/-- The xor of two bitvectors. -/
+protected def xor (x y : BitVec w) : BitVec w :=
+  ⟨x.val ^^^ y.val, by simp [HXor.hXor, Xor.xor, Nat.xor, Nat.bitwise_lt]⟩
+
+/-- The left-shift of a bitvector by some amount. -/
+protected def shiftLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  .ofNat w (x.val <<< n)
+
+/-- The right-shift of a bitvector by some amount. -/
+protected def shiftRight (x : BitVec w) (n : Nat) : BitVec w :=
+  ⟨x.val >>> n, by
+      simp only [Nat.shiftRight_eq_div_pow]
+      exact lt_of_le_of_lt (Nat.div_le_self' _ _) (x.isLt) ⟩
```
Either you should remove these definitions (and inline their implementation into the `instance` below), or you should add a simp lemma for each that replaces it with the corresponding notation.

mhk119 commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,189 @@
+-/
+/-- The mod of two bitvectors. -/
+protected def mod (x y : BitVec w) : BitVec w :=
+  ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.isLt⟩
```
Now that ```BitVec w``` is an abbrev over ```Fin 2^w```, I don't think we can change that.

mhk119 commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,189 @@
+  ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.isLt⟩
+
+/-- Dividing two bitvectors. -/
+protected def div (x y : BitVec w) : BitVec w :=
```
Since ```BitVec w``` is now an ```abbrev Fin 2^w```, I think all of the definitions that required it are either gone or they now mention which one is used.

alexkeizer commented:
Now that #7410 is merged to master, I merged master back into this branch and removed the previous proof of equality between `bitwise` and `bitwise'`

eric-wieser commented:
I'd be somewhat in favor of wrapping up #7451 before coming back to this, if you're ok with that.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,119 @@
-def toFin {n : ℕ} (i : Bitvec n) : Fin (2 ^ n) :=
-  i.toNat
-#align bitvec.to_fin Bitvec.toFin
+instance : Complement (BitVec w) := ⟨(· + (1 : BitVec w))⟩
```
This looks wrong to me; should't it be `-1 - x`, so that `0b110 = 6` is sent to `0b001 = 1 = -7`?

eric-wieser replied:
(Or better, `finReverse` or whatever the existing map is called)

semorrison commented:
Leo would like https://github.com/leanprover/std4/pull/286 merged soon, as it is needed for other projects.

I think the two options are:
* just wait until that PR lands, and then adapt this PR
* make whatever suggestions you like on that PR, but hopefully quite quickly! Please don't add lemmas to the Std library beyond basic simp lemmas or lemmas characterising operations. (Or perhaps simpler: let's just not add any for now.)

eric-wieser commented:
I'm rather confused about how this code overlaps with that code. Where did Leo's code come from; did you develop this PR independently, @alexkeizer? Are both PRs copied from some work by Joe Hendrix that is publicly available?

mhk119 commented:
> I'm rather confused about how this code overlaps with that code. Where did Leo's code come from; did you develop this PR independently, @alexkeizer? Are both PRs copied from some work by Joe Hendrix that is publicly available?

I started working on the PR that Joe Hendrix started [here](https://github.com/ufmg-smite/lean-smt/blob/main/Smt/Data/BitVec.lean). Perhaps Leo has the same file?

mhk119 commented:
> Leo would like [leanprover/std4#286](https://github.com/leanprover/std4/pull/286) merged soon, as it is needed for other projects.
> 
> I think the two options are:
> 
> * just wait until that PR lands, and then adapt this PR
> * make whatever suggestions you like on that PR, but hopefully quite quickly! Please don't add lemmas to the Std library beyond basic simp lemmas or lemmas characterising operations. (Or perhaps simpler: let's just not add any for now.)

mhk119 commented:
> Leo would like [leanprover/std4#286](https://github.com/leanprover/std4/pull/286) merged soon, as it is needed for other projects.
> 
> I think the two options are:
> 
> * just wait until that PR lands, and then adapt this PR
> * make whatever suggestions you like on that PR, but hopefully quite quickly! Please don't add lemmas to the Std library beyond basic simp lemmas or lemmas characterising operations. (Or perhaps simpler: let's just not add any for now.)

Thanks! This is exciting. I believe having these BitVec definitions in Std is a great step forward. I'm in favor of going with option 1.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,56 @@
-@[reducible]
-def Bitvec (n : ℕ) :=
-  Vector Bool n
-#align bitvec Bitvec
```
Can you try to keep these aligns? For instance, this one would become `#align bitvec Std.BitVec`

alexkeizer commented:
I've now refactored this PR to include the changes made in #7451 and https://github.com/leanprover/std4/pull/286. This means a rather big part of this PR has been removed (since it was upstreamed in Std)

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,102 @@
-It is not fully in compliance with mathlib style standards.
+/-!
+# Basic operations on bitvectors
+We define bitvectors. We choose the `Fin` representation over others for its relative efficiency
```
I guess this docstring is no longer accurate (and nor is the PR title)

alexkeizer replied:
I've updated both

alexkeizer commented:
Sorry about that. The proof broke, I think because the definition of `getLsb` changed. I wanted to get the CI green first, to make sure no downstream files got broken, so I removed it

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,84 @@
+  ⟨(match x, y, · with | ⟨_, _⟩,⟨_, _⟩, rfl => rfl), (· ▸ rfl)⟩
+
+/-- `x < y` as natural numbers if and only if `x < y` as `BitVec w`. -/
+theorem toNat_lt_toNat {x y : BitVec w} : (x.toNat < y.toNat) = (x < y) := rfl
```
```suggestion
theorem toNat_lt_toNat {x y : BitVec w} : x.toNat < y.toNat ↔ x < y := rfl
```

mhk119 commented:
> Sorry about that. The proof broke, I think because the definition of `getLsb` changed. I wanted to get the CI green first, to make sure no downstream files got broken, so I removed it

Ok cool. I proved the sorry using another outline. Do you still want those three `sorry`-ed lemmas? I removed them for now because I thought it was just for filling in that `sorry` in `get_eq_testBit`.

alexkeizer commented:
> > Sorry about that. The proof broke, I think because the definition of `getLsb` changed. I wanted to get the CI green first, to make sure no downstream files got broken, so I removed it
> 
> Ok cool. I proved the sorry using another outline. Do you still want those three `sorry`-ed lemmas? I removed them for now because I thought it was just for filling in that `sorry` in `get_eq_testBit`.

That's perfectly fine, there's no need for those lemmas. It's better to go with the phrasing in terms of multiplication and power, since those are the simp normal forms, anyway. I just did some cleanup, moving the lemma about `Nat`s to the `Nat.Bitwise` file.


I noticed that the files are currently under a `Bitvec` folder, while the type is `BitVec` (with capital `V`). Do we want to change that, to make imports more consistent?

alexkeizer commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -0,0 +1,74 @@
+  simp [BitVec.ofNat]
+  apply Fin.cast_val_eq_self x
+
+lemma ofNat_toNat' (x : BitVec w) (h : v = w):
+    HEq x (BitVec.ofNat v x.toNat) := h ▸ heq_of_eq (ofNat_toNat x).symm
```
```suggestion@[simp]
lemma cast_eq (x : BitVec w) (h : w = w) : x.cast h = x :=
  rfl

lemma ofNat_toNat' (x : BitVec w) (h : w = v):
    BitVec.ofNat v x.toNat = x.cast h := by
  cases h; rw [ofNat_toNat, cast_eq]
```

Heterogeneous equalities are a pain to work with, I would suggest using `BitVec.cast` here

digama0 commented:
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

eric-wieser commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -52,8 +52,13 @@ lemma ofNat_toNat (x : BitVec w) : BitVec.ofNat w x.toNat = x := by
-lemma ofNat_toNat' (x : BitVec w) (h : v = w):
-    HEq x (BitVec.ofNat v x.toNat) := h ▸ heq_of_eq (ofNat_toNat x).symm
+@[simp]
+lemma cast_eq (x : BitVec w) (h : w = w) : x.cast h = x :=
+  rfl
```
```suggestion
lemma cast_eq (x : BitVec w) : x.cast rfl = x :=
  rfl
```

eric-wieser commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -0,0 +1,79 @@
+  simp only [HAppend.hAppend, append, toNat_ofFin]
+  rw [toNat_ofNat (Nat.add_comm w v ▸ append_lt hl hm)]
+
+theorem toNat_extractLsb {i j} {x : BitVec w} :
+    (extractLsb i j x).toNat = x.toNat / 2 ^ j % (2 ^ (i - j + 1)) := by
+  simp [extractLsb, extractLsb', shiftRight_eq_div_pow]
```
I would guess that Mathlib doesn't want to both writing lemmas about `extractLsb` because it's awkward to use; I think we should just have a simp lemma that rewrites `extractLsb` as `extractLsb'`, and write all the lemmas about that.

alexkeizer replied:
I've added 
```lean
@[simp]
lemma extractLsb_eq {w : ℕ} (hi lo : ℕ) (a : BitVec w) :
    extractLsb hi lo a = extractLsb' lo (hi - lo + 1) a :=
  rfl
```

ChrisHughes24 commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +0,0 @@
---  simp [bits_to_nat_to_list]
---  unfold bits_to_nat add_lsb List.foldl cond
---  simp [cond_to_bool_mod_two]
-theorem bits_toNat_decide (n : ℕ) : Bitvec.toNat (decide (n % 2 = 1) ::ᵥ Vector.nil) = n % 2 := by
```
This appears to have been deleted

tobiasgrosser replied:
This seems to be a helper theorem that was only used in the old `to_nat_of_nat`. I am wondering, how would you even state it on the current BitVec representation?

alexkeizer replied:
We can define a `cons` operation as roughly `cons false x = x` and `cons true x = x + 2^n`, but I don't think this result would be particularly interesting

eric-wieser replied:
I think deleting this is fine, but you should never delete an `#align`; instead, replace it with `noalign`.

tobiasgrosser replied:
Thank you, Eric. I added #noalign for bitvec.bits_to_nat_to_bool and bitvec.of_nat_succ.

alexkeizer commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -0,0 +1,84 @@
+  simp only [HAppend.hAppend, append, toNat_ofFin]
+  rw [toNat_ofNat (Nat.add_comm w v ▸ append_lt hl hm)]
+
+@[simp]
```
This was suggested by @eric-wieser 
> I would guess that Mathlib doesn't want to both writing lemmas about `extractLsb` because it's awkward to use; I think we should just have a simp lemma that rewrites `extractLsb` as `extractLsb'`, and write all the lemmas about that.

(BTW, when making a review comment about a piece of code that spans multiple lines, you can click-and-drag, that way you can control which lines GitHub will show. Currently it only shows the `@[simp]` attribute and a few preceding, unrelated lines, making it a bit harder to see that it is about
```lean
@[simp]
lemma extractLsb_eq {w : ℕ} (hi lo : ℕ) (a : BitVec w) :
    extractLsb hi lo a = extractLsb' lo (hi - lo + 1) a :=
  rfl
```
)

alexkeizer commented on Mathlib/Data/BitVec/Defs.lean:
```diff
@@ -0,0 +1,104 @@
+#align bitvec.to_int Std.BitVec.toInt
+
+#noalign bitvec.one
+#noalign bitvec.adc
+#noalign bitvec.sbb
+#noalign bitvec.uborrow
+#noalign bitvec.sborrow
```
Sure, but would it be OK to do this in another PR? I currently have a second PR ready that gives shows that `BitVec` is a commutative ring, but it's blocked on the current PR. If we merge this one, then we can work on bringing back these definitions in parallel to the Ring-PR being reviewed

ChrisHughes24 replied:
I think the convention is that mathlib never gets worse. We shouldn't lose things and get them back, just in case somebody needs them and they get annoyed.

alexkeizer replied:
Turns out that `sborrow` and `uborrow` are just `slt` and `ult`, so I aligned them to that.
I've also added `adc` and `sbb`, although I would personally classify those as auxiliary defs, and not particularly interesting by themselves. For context, this is their implementation
```lean
/-- Add with carry (no overflow) -/
def adc {n} (x y : BitVec n) (c : Bool) : BitVec (n+1) :=
  ofFin (x.toNat + y.toNat + c.toNat)
#align bitvec.adc Std.BitVec.adc

/-- Subtract with borrow -/
def sbb {n} (x y : BitVec n) (b : Bool) : Bool × BitVec n :=
  let y := y + ofFin b.toNat
  (x < y, x - y)
#align bitvec.sbb Std.BitVec.sbb
```

mhk119 commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -326,6 +344,55 @@ theorem bitwise_swap {f : Bool → Bool → Bool} :
+It is almost always specialized with `z = 0` and `i = w`; the length of the binary representation.
+Note that `ofBits f z i = 2 ^ i * z + ofBits f 0 i` which we prove next.
+This is an alternative to using `List`. It will be used for bitadd, bitneg, bitmul etc.-/
+def ofBits (f : Nat → Bool) (z : Nat) : Nat → Nat
```
Because `z` is necessarily needed in the definition. At each step, it gets populated by the carry-out bit and then gets appended to the binary representation.

alexkeizer replied:
How about
```lean
def ofBits (f : Nat → Bool) : Nat → Nat :=
  go 0
  where
    go (z : Nat) : Nat → Nat
    | 0 => z
    | i + 1 => go (z.bit (f i)) i
```
This conveys properly that `z` is just an internal accumulator, and that the public API should only ever have `z = 0`

tobiasgrosser replied:
I moved the proves to this version of ofBits.

mhk119 replied:
Thanks!

eric-wieser commented on Mathlib/Data/BitVec/Defs.lean:
For the purpose of a nicer diff, can you rename this back to `Bitvec` with a lowercase `v`? You can then rename it back to `V` in a follow-up PR.

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -326,6 +344,61 @@ theorem bitwise_swap {f : Bool → Bool → Bool} :
+It takes a boolean function `f` on each bit the number of bits `i`.  It is
+almost always specialized  `i = w`; the length of the binary representation.
+This is an alternative to using `List`. It will be used for bitadd, bitneg, bitmul etc.-/
+def ofBits (f : Nat → Bool) : Nat → Nat :=
```
I'd expect this to be more convenient as 
```suggestion
def ofBits {i : Nat} (f : Fin n → Bool) : Nat :=
```
that is, make `f` dependently typed. Otherwise you need a congruence lemma that says `f j` is irrelevant unless `j < i`.

tobiasgrosser replied:
I tried but don't yet see what to do here. Someone else has an idea?

eric-wieser replied:
I mean something like
```
def ofBits {n : Nat} (f : Fin n → Bool) : Nat :=
  go f 0
  where
    /-- A helper method where `z` the starting point.
    Note that `ofBits.go f z = 2 ^ n * z + ofBits f i` which we prove next. -/
    go {n : Nat} (f : Fin n → Bool) (z : Nat) : Nat :=
    match n with
    | 0 => z
    | _ + 1 => go (f ∘ Fin.castSucc) (z.bit (f (.last _)))
```
I'll push something to that effect shortly.

mhk119 replied:
What is the motivation behind defining it in terms of `Fin`? I originally defined it this way to avoid lists (which is equivalent to this new definition). This was mainly because our proofs of bit-blasting theorems all follow this strong induction format where we descend to smaller binary representations (removing elements is painful to do with lists).

Also, I'm trying to prove `testBit_ofBits` with this new definition (which was a simple 5-line strong inductive proof before) but I can't seem to generalize over `i` anymore. Issues like this make me favor the `Nat` definition over the `Fin` definition (dependent types are harder to work with). Am I missing something?

alexkeizer replied:
In some sense, the `Fin` design is more precise, since it's now obvious that only the values of `f i` where `i < n` matter. The previous design required garbage values `f j` for all `j >= n`, and, like Eric mentioned, some congruence lemma stating that these garbage values don't influence the output.

That said,
> I'd expect this to be more convenient
So far, the proofs that were broken are considerable less convenient. 
I've tried to play around with different `Fin` recursion principles to regain the proofs as straightforward inductions, but have not gotten that to work.

@eric-wieser: What do you think of keeping the old `Nat -> Bool`-based version as a (private) implementation detail, and have the `Fin n -> Bool` version just wrap that one (by extending the function with canonical garbage values). That should be easily implemented over the previously working version.

tobiasgrosser replied:
I like that idea.

tobiasgrosser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,152 @@
-
-theorem toNat_eq_foldr_reverse {n : ℕ} (v : Bitvec n) :
-    v.toNat = v.toList.reverse.foldr (flip addLsb) 0 := by rw [List.foldr_reverse]; rfl
-#align bitvec.to_nat_eq_foldr_reverse Bitvec.toNat_eq_foldr_reverse
```
@eric-wieser, do you think this lemma is needed? Its a little hard for me to prove, but maybe someone else could try? Or we leave it out if not strictly necessary.

tobiasgrosser replied:
I guess the reverse in the theorem name does not make sense anymore. I wonder if bits maybe returned the bits in the opposite order than toList?

alexkeizer replied:
Indeed, the old `toList` returned the bits in big-endian order (where the head of the list is the MSB), whereas `Nat.bits` returns bits in little-endian order (the head of the list is the LSB).

tobiasgrosser replied:
Should we then drop the reverse from the theorem name?

alexkeizer replied:
I think this theorem as it is currently stated is quite different from the original.
The current statement:
```lean
v.toNat = v.toNat.bits.foldr bit 0
```
is really a theorem about `Nat`s, not `BitVec`s.

The equivalent would be something akin to
```lean
v.toNat = v.toBEList.reverse.foldr (flip addLsb) 0
```
Although this should be equivalent to  
```lean
v.toNat = v.toBEList.foldl (flip addLsb) 0
```

I'm not sure why the original theorem used `reverse` and `foldr` over `foldl`

alexkeizer replied:
An update on this: to prove `v.toNat = v.toBEList.foldl (flip addLsb) 0` cleanly, we really want to be able to do induction on a bitvector in terms of a `cons` or `concat` operation. Adding this in the current PR seems undesirable, since that would make the PR even larger.
So, instead I propose we change course, rather than replace the entire `Bitvec` API in one go, let's keep both `Bitvec` and `Std.BitVec` types while incrementally adding results to the latter. I've opened #8345 as a first step in that direction.

tobiasgrosser replied:
I removed this lemma for now and replaced it with #noalign in this PR. We can then decide if we want to merge this PR with noalign or instead go for #8345.

eric-wieser replied:
I think merging with noaligns is fine

eric-wieser commented on Mathlib/Data/Bool/Basic.lean:
In the interest of slowly shrinking this PR (and saving you more expensive rebuilds), can you split the changes in this file to their own PR, which I'll merge quickly?

alexkeizer replied:
Done, in #8328

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,160 @@
-theorem bitsToNat_toList {n : ℕ} (x : Bitvec n) : Bitvec.toNat x = bitsToNat (Vector.toList x) :=
+variable {w v : Nat}
+
+/-! We add simp-lemmas that rewrite bitvector operations into the equivalent notation -/
+@[simp] lemma append_eq (x : BitVec w) (y : BitVec v)   : BitVec.append x y = x ++ y        := rfl
+@[simp] lemma shiftLeft_eq (x : BitVec w) (n: Nat)      : BitVec.shiftLeft x n = x <<< n    := rfl
+@[simp] lemma ushiftRight_eq (x : BitVec w) (n: Nat)    : BitVec.ushiftRight x n = x >>> n  := rfl
+@[simp] lemma not_eq (x : BitVec w)                     : BitVec.not x = ~~~x               := rfl
+@[simp] lemma and_eq (x y : BitVec w)                   : BitVec.and x y = x &&& y          := rfl
+@[simp] lemma or_eq (x y : BitVec w)                    : BitVec.or x y = x ||| y           := rfl
+@[simp] lemma xor_eq (x y : BitVec w)                   : BitVec.xor x y = x ^^^ y          := rfl
+@[simp] lemma neg_eq (x : BitVec w)                     : BitVec.neg x = -x                 := rfl
+@[simp] lemma add_eq (x y : BitVec w)                   : BitVec.add x y = x + y            := rfl
+@[simp] lemma sub_eq (x y : BitVec w)                   : BitVec.sub x y = x - y            := rfl
+@[simp] lemma mul_eq (x y : BitVec w)                   : BitVec.mul x y = x * y            := rfl
```
I think you could PR these ones straight to Std, since they're fundamentally about deciding what's the API and what's an implementation detail.

alexkeizer replied:
leanprover/std4#357

alexkeizer commented:
It seems this PR is mostly blocked on not wanting to lose existing results about bitvectors. Replicating all these results for `Std.BitVec` in a single PR is leading to a rather oversized PR, so I've opened #8345 with an alternative approach: just keep the existing `Bitvec` type and its result, and develop the `Std.BitVec` API in parallel.

alexkeizer commented:
I've removed some new theorems from the `Nat.Bitwise` file which are not needed for the main `BitVec` contributions.
Since this PR does not depend on these removed results, I've not opened a PR for them, but I have saved them in a branch: mathlib4:textbit-lemmas

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -338,6 +347,25 @@ theorem bitwise_comm {f : Bool → Bool → Bool} (hf : ∀ b b', f b b' = f b'
 
+/-- If `x` and `y` fit within `n` bits, then the result of any (well-behaved) bitwise operation on
+    `x` and `y` also fits within `n` bits -/
+theorem bitwise_lt {f x y n} (hx : x < 2 ^ n) (hy: y < 2 ^ n) (h: f false false = false) :
+    bitwise f x y < 2 ^ n := by
```
Is `h` needed here?

alexkeizer replied:
Good catch, I think the result should still be true even without `h`.
That said, the current proof does rely on `h`.

eric-wieser replied:
Fixed. I think the answer is always `binaryRec'`

tobiasgrosser replied:
Thank you @eric-wieser. This is much appreciated.

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -338,6 +347,25 @@ theorem bitwise_comm {f : Bool → Bool → Bool} (hf : ∀ b b', f b b' = f b'
 
+/-- If `x` and `y` fit within `n` bits, then the result of any (well-behaved) bitwise operation on
+    `x` and `y` also fits within `n` bits -/
+theorem bitwise_lt {f x y n} (hx : x < 2 ^ n) (hy: y < 2 ^ n) (h: f false false = false) :
```
```suggestion
theorem bitwise_lt {f x y n} (hx : x < 2 ^ n) (hy: y < 2 ^ n) (h : f false false = false) :
```

tobiasgrosser commented:
I removed another round of likely unrelated changes to the Bitwise files.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,143 @@
+def toLEList (x : BitVec w) : List Bool :=
+  List.ofFn x.getLsb'
 
-/-- Return the integer encoded by the input bitvector -/
-protected def toInt : ∀ {n : Nat}, Bitvec n → Int
-  | 0, _ => 0
-  | succ _, v =>
-    cond (head v) (Int.negSucc <| Bitvec.toNat <| Bitvec.not <| tail v)
-      (Int.ofNat <| Bitvec.toNat <| tail v)
-#align bitvec.to_int Bitvec.toInt
```
Where did this go?

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,143 @@
+* `Std.BitVec.sge`: Signed greater-equals comparison of bitvectors
+* `Std.BitVec.ugt`: Unsigned greater-than comparison of bitvectors
+* `Std.BitVec.uge`: Unsigned greater-equals comparison of bitvectors
+* `Std.BitVec.ofLEList`: Turn a little-endian list of Booleans into a bitvector
+* `Std.BitVec.ofBEList`: Turn a big-endian list of Booleans into a bitvector
```
I don't see these below in the file.

tobiasgrosser replied:
These were referring functions that we decided to not upstream in the initial PR. They are now removed.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,143 @@
 /-
-Copyright (c) 2015 Joe Hendrix. All rights reserved.
+Copyright (c) 2023 Wojciech Nawrocki. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Joe Hendrix, Sebastian Ullrich
+Authors: Harun Khan, Alex Keizer, Abdalrhman M Mohamed
 -/
```
Why does the copyright mention a non-author?

tobiasgrosser replied:
This code is an import from @mhk119 and his collegues at Stanford. AFAIU Abdalrhman co-authored some of the pieces, but is just not involved in upstreaming. @mhk119, can you clarify?

alexkeizer replied:
I think Eric is referring to the fact that Wojciech Nawrocki is listed as the copyright-holder, but not as an author. I think that's because the original PR adapted code by Wojciech in a different project. However, I suspect that all of this person's code has now been upstreamed to Std, so the copyright attribution might no longer make sense for this PR in the current form

eric-wieser replied:
Yes, exactly; it's @Vtec234's mention in the copyright but not the authors that confuses me.

tobiasgrosser replied:
I see. So I guess we either add Wojciech to the authors or add somebody else to the copyright. @mhk119, what would you suggest?

eric-wieser replied:
Frankly I would just leave the original 2015 copyright and *maybe* add the three new authors. Given the file is mostly aligns, I don't think there's much to claim authorship of anyway.

tobiasgrosser replied:
That is a good idea. I changed the headers accordingly.

mhk119 replied:
Hi, sorry for the late reply. Daytime in the UK is nighttime in California :). Sure, @eric-wieser's suggestion sounds good. Thanks!

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -316,6 +319,12 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
     simp [h]
 #align nat.test_bit_two_pow Nat.testBit_two_pow
 
+lemma and_two_pow {n i} : n &&& 2 ^ i = (n.testBit i).toNat * 2 ^ i := by
+  apply eq_of_testBit_eq; intro j
+  rw [mul_comm, testBit_land]
+  cases' h : n.testBit i <;> cases' (ne_or_eq i j) with h1 h1
+  <;> simp [testBit_two_pow_of_ne _, *] at * <;> assumption
```
Just noting this is currently under review in another open PR, though I forget which one.

tobiasgrosser replied:
It is under review in https://github.com/leanprover-community/mathlib4/pull/8558.

tobiasgrosser replied:
This has now been merged to master and this PR does not need to ad it.

eric-wieser commented on Mathlib/Data/Bitvec/Defs.lean:
```diff
@@ -1,339 +1,141 @@
 
-instance : Sub (Bitvec n) :=
-  ⟨Bitvec.sub⟩
+/-- Unsigned greater than for bitvectors. -/
+protected def ugt (x y : BitVec w) : Bool := BitVec.ult y x
+#align bitvec.ugt Std.BitVec.ugt
```
What's the rationale for changing this and the declarations around it from `Prop` to `bool`? Can you add an explanation in the PR description?

alexkeizer replied:
The choice was made to be consistent with Std, I believe also with input from an earlier review by @ChrisHughes24. I've edited the PR description to clarify the motivation behind this change

eric-wieser commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -341,6 +341,35 @@ theorem bitwise_comm {f : Bool → Bool → Bool} (hf : ∀ b b', f b b' = f b'
+lemma shiftLeft_lt {x n m : ℕ} (h : x < 2 ^ n) : x <<< m < 2 ^ (n + m) := by
+  simp only [pow_add, shiftLeft_eq, mul_lt_mul_right (two_pow_pos _), h]
+
+lemma append_lt {x y n m} (hx : x < 2 ^ n) (hy : y < 2 ^ m) : y <<< n ||| x < 2 ^ (n + m) := by
```
I think this could do with a docstring expaining that this expression happens to coincide with the implementation of `Std.BitVec.append`.

In the longer term, it would make sense to upstream this to avoid an unecessary modulo operation in `append`; but that's definitely out of scope here.

alexkeizer replied:
This has been resolved in #8598

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,115 @@
+  simp [addLsb, two_mul]; cases b <;> rfl
+#align bitvec.add_lsb_eq_twice_add_one Std.BitVec.addLsb_eq_twice_add_one
+
+#noalign bitvec.to_nat_eq_foldr_reverse
```
Let's leave a breadcrumb
```suggestion
-- The previous statement was `(v : Bitvec n) : v.toNat = v.toList.reverse.foldr (flip addLsb) 0`.
-- Since the statement is awkward and `Std.BitVec` has no comparable API, we just drop it.
#noalign bitvec.to_nat_eq_foldr_reverse
```

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,115 @@
-    simp [Nat.mul_succ]
-  · simp
-    apply xs_ih
-#align bitvec.to_nat_append Bitvec.toNat_append
```
This align seems to have gone; I think it's a `noalign` since it's a very different statement to the new one, unless you want to implement `toNat_concat` (did `concat` land in Std?)

alexkeizer replied:
`concat` did land in Std, but I don't think it's particularly useful to implement `toNat_concat`, let's just go with `noalign`

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,115 @@
+lemma extractLsb_eq {w : ℕ} (hi lo : ℕ) (a : BitVec w) :
+    extractLsb hi lo a = extractLsb' lo (hi - lo + 1) a :=
   rfl
-#align bitvec.of_nat_succ Bitvec.ofNat_succ
```
This should be a `noalign`

eric-wieser replied:
(or should be reinstated with `cons` / `concat`)

alexkeizer replied:
I went with `noalign`, leaving a comment with what the statement in terms of `concat` would be

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,124 @@
+
+/-- `x < y` as natural numbers if and only if `x < y` as `BitVec w`. -/
+theorem toNat_lt_toNat {x y : BitVec w} : x.toNat < y.toNat ↔ x < y :=
+  ⟨id, id⟩
```
```suggestion
  Iff.rfl
```

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,124 @@
+
+variable {w v : Nat}
+
+theorem toNat_inj {x y : BitVec w} : x.toNat = y.toNat ↔ x = y :=
```
Can you prove the statement with `Injective`, then prove this with `.eq_iff`?

tobiasgrosser replied:
I gave this a shot. (Btw, thank you for helping on zulip).

I guess my proof likely can be golfed and I am also unsure if the statement using `@` is sensible.

tobiasgrosser replied:
This has now be golved.

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -21,8 +21,17 @@ open Nat
 variable {w v : Nat}
 
-theorem toNat_inj {x y : BitVec w} : x.toNat = y.toNat ↔ x = y :=
-  ⟨(match x, y, · with | ⟨_, _⟩,⟨_, _⟩, rfl => rfl), (· ▸ rfl)⟩
```
You could have just kept the first half of this proof :)

tobiasgrosser replied:
I gave it another shot, but I am afraid it is not yet minimal. Advice is much appreciated, otherwise we will try to reduce it further on Monday.

tobiasgrosser commented:
I updated master. There are no no changes to Nat/Bitwise as the last ones were merged in #8598.

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -1,173 +1,131 @@
+
+variable {w v : Nat}
+
+theorem toNat_inj {n : Nat}: Function.Injective (@Std.BitVec.toNat n) := by
```
I think my comment earlier got lost to GitHub; you should be able to prove this with a substring of the old proof (in one line), rename it to `toNat_injective`, then prove the previous `toNat_inj` in terms of the `Injective` statement as we do elsewhere in mathlib.

tobiasgrosser replied:
Alex helped to clean this up.

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -21,15 +21,10 @@ open Nat
 
-theorem toNat_inj {n : Nat}: Function.Injective (@Std.BitVec.toNat n) := by
+theorem toNat_injective {n : Nat}: Function.Injective (@Std.BitVec.toNat n) := by
   unfold BitVec.toNat Function.Injective toFin
```
I think you can drop this line entirely

tobiasgrosser replied:
Alex helped clean this up

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -21,10 +21,11 @@ open Nat
+theorem toNat_injective {n : Nat}: Function.Injective (@Std.BitVec.toNat n)
+  | ⟨_, _⟩, ⟨_, _⟩, rfl => rfl
+
+theorem toNat_inj {x y : BitVec w} : x.toNat = y.toNat → x = y :=
```
```suggestion
theorem toNat_inj {x y : BitVec w} : x.toNat = y.toNat ↔ x = y :=
```

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -21,10 +21,11 @@ open Nat
+  | ⟨_, _⟩, ⟨_, _⟩, rfl => rfl
+
+theorem toNat_inj {x y : BitVec w} : x.toNat = y.toNat → x = y :=
+  @toNat_injective w x y
```
```suggestion
  toNat_injective.eq_iff
```

eric-wieser commented on Mathlib/Data/Bitvec/Lemmas.lean:
```diff
@@ -21,10 +21,11 @@ open Nat
-  intros x y
-  exact (match x, y, · with | ⟨_, _⟩,⟨_, _⟩, rfl => rfl)
+theorem toNat_injective {n : Nat}: Function.Injective (@Std.BitVec.toNat n)
+  | ⟨_, _⟩, ⟨_, _⟩, rfl => rfl
```
Nice!

tobiasgrosser replied:
Indeed. 👍

eric-wieser commented:
bors merge

This looks good to go now; thanks for splitting out all the other smaller PRs to make this a manageable size!

tobiasgrosser commented:
Thank you all, indeed!

mhk119 commented:
Thanks everyone for your hard work! This is amazing.

linesthatinterlace commented:
Congratulations!
